### PR TITLE
Fixes #22687 - Add openscap report check and procedure

### PR DESCRIPTION
**Pull request by xprazak2**
wip as I need to test in production

### CODE REVIEW

File: definitions/checks/foreman_openscap/report_associations.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+module Checks
+  module ForemanOpenscap
+    class ReportAssociations < ForemanMaintain::Check
+      metadata do
+        label :report_associations
```
Comment (iNecas) [inline definitions/checks/foreman_openscap/report_associations.rb]: I would suggest using `:openscap_` prefix in the label: it makes it easier when listing the checks in `fm health list` command

File: definitions/checks/foreman_openscap/report_associations.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        for_feature :foreman_openscap
+        description 'Check whether reports have correct associations'
+        tags :pre_upgrade, :foreman_openscap, :report_associations
+
+        confine do
```
Comment (iNecas) [inline definitions/checks/foreman_openscap/report_associations.rb]: When using `for_feature` the `confine` is not needed

File: definitions/checks/foreman_openscap/report_associations.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module Checks
+  module ForemanOpenscap
+    class ReportAssociations < ForemanMaintain::Check
```
Comment (iNecas) [inline definitions/checks/foreman_openscap/report_associations.rb]: The check should describe what's wrong about the state, if it fails: what about something like `InvalidReportAssociations`

File: definitions/procedures/foreman_openscap/report_associations.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
+module Procedures::ForemanOpenscap
+  class ReportAssociations < ForemanMaintain::Procedure
```
Comment (iNecas) [inline definitions/procedures/foreman_openscap/report_associations.rb]: Similarly to the checks naming, I would suggest `ClearReportAssociations`

**Comment by ntkathole** [comment]
```
 ./bin/foreman-maintain advanced procedure run foreman-openscap-report-associations
Running ForemanMaintain::Scenario
================================================================================
Delete reports with association issues:                               [FAIL]
undefined method `join' for nil:NilClass
--------------------------------------------------------------------------------
Scenario [ForemanMaintain::Scenario] failed.

The following steps ended up in failing state:

  [foreman-openscap-report-associations]

Resolve the failed steps and rerun
the command. In case the failures are false positives,
use --whitelist="foreman-openscap-report-associations"
```

**Comment by xprazak2** [comment]
I made changes as suggested. I am not entirely sure what causes the nil issue as I cannot reproduce, but I made an extra precaution for that as well.

**Comment by ntkathole** [comment]
@xprazak2 I still see the same issue o sat62 and sat63. Let me know if you need setup.
```
E, [2018-03-18 03:39:41-0400 #25968] ERROR -- : undefined method `join' for nil:NilClass (NoMethodError)
/root/foreman_maintain/definitions/features/foreman_openscap.rb:37:in `delete_reports'
/root/foreman_maintain/definitions/procedures/foreman_openscap/invalid_report_associations.rb:12:in `run'
/root/foreman_maintain/lib/foreman_maintain/executable.rb:119:in `__run__'
/root/foreman_maintain/lib/foreman_maintain/runner/execution.rb:76:in `block (2 levels) in run'
/root/foreman_maintain/lib/foreman_maintain/runner/execution.rb:99:in `capture_errors'
/root/foreman_maintain/lib/foreman_maintain/runner/execution.rb:75:in `block in run'
/root/foreman_maintain/lib/foreman_maintain/runner/execution.rb:93:in `with_metadata_calculation'
/root/foreman_maintain/lib/foreman_maintain/runner/execution.rb:74:in `run'
/root/foreman_maintain/lib/foreman_maintain/runner.rb:109:in `run_step'
/root/foreman_maintain/lib/foreman_maintain/runner.rb:98:in `run_steps'
/root/foreman_maintain/lib/foreman_maintain/runner.rb:91:in `execute_scenario_steps'
/root/foreman_maintain/lib/foreman_maintain/runner.rb:45:in `run_scenario'
```

File: definitions/procedures/foreman_openscap/invalid_report_associations.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module Procedures::ForemanOpenscap
+  class InvalidReportAssociations < ForemanMaintain::Procedure
+    metadata do
+      param :ids_to_remove, 'Ids of reports to remove'
```
Comment (kgaikwad) [inline definitions/procedures/foreman_openscap/invalid_report_associations.rb]: @xprazak2 ,

param ` :ids_to_remove` is required to run this procedure so you could mention `:required => true` to param declaration.

**Comment by xprazak2** [comment]
The nil issue was caused by running the procedure directly through `advanced procedure run`, but it requires a list of ids which are supplied by check. I added `advanced_run false` to disable running it directly. Thanks @kgaikwad!

**Comment by ntkathole** [comment]
Tested on sat 6.2 : 
Steps:
1. Created scap report with policy
2. deleted policy

Before performing upgrade, it asks to delete orphaned reports. But my concern is if we are handling the same in installer as well https://bugzilla.redhat.com/show_bug.cgi?id=1547607#c16 , @xprazak2 @iNecas shall we still have this check in foreman-maintain ?

**Comment by xprazak2** [comment]
Partially, the migration in bz just removes reports without policy. Here we check for other associations as well. The migration runs as a part of upgrade and only once, having this in foreman_maintain provides value to users outside of upgrade scenario.

**Comment by xprazak2** [comment]
@iNecas, any additional comments?

**Comment by iNecas** [comment]
Last question: should this apply to any upgrade with 'tfm-rubygem-foreman_openscap' >= '0.5.3', or is this specific let's say for sat6.3 -> sat6.4 but would not be applicable for sat6.4 -> sat6.5?

**Comment by xprazak2** [comment]
People were having problems on 6.2, so this is mostly relevant for 6.2 -> 6.3. I do not expect this problem in later versions, steps were taken to prevent this from happening again. So not too relevant for the 6.3 -> 6.4 and later, but the check and procedure can be run manually as well if the users need it.

**Comment by iNecas** [comment]
Ok, let's pull it in then. Thanks @xprazak2

**Pull request merged**